src/compiler/checker.ts

```diff
@@ -4436,17 +4517,20 @@ namespace ts {
                 // it may hold in a structural comparison.
                 // Report structural errors only if we haven't reported any errors yet
                 let reportStructuralErrors = reportErrors && errorInfo === saveErrorInfo;
-                // identity relation does not use apparent type
+                // Identity relation does not use apparent type
                 let sourceOrApparentType = relation === identityRelation ? source : getApparentType(source);
-                if (sourceOrApparentType.flags & TypeFlags.ObjectType && target.flags & TypeFlags.ObjectType) {
+                // In a check of the form X = A & B, we will have previously checked if A relates to X or B relates
+                // to X. Failing both of those we want to check if the aggregation of A and B's members structurally
+                // relates to X. Thus, we include intersection types on the source side here.
+                if (sourceOrApparentType.flags & (TypeFlags.ObjectType | TypeFlags.Intersection) && target.flags & TypeFlags.ObjectType) {

```

Comment:
I think the real issue here is that optional properties cannot be checked in the face of unknown extensions. We allow assignment to a target with an optional property when the source doesn't have that property, but we only know that to be true when the source is not subject to further extension (i.e. when source type expresses "the whole truth"). For that reason, we do have this same problem with type parameters. For example, X = A & T, where X is a type with an optional property p, A is a type that has no property p, and T is a type parameter. If T is instantiated to a type that has a property p with the wrong type, an inconsistency appears. In fact, the same inconsistency exists just with X = T. It all has to do with the way we check optional properties and I'm not sure we can do much about it.